File: frontend/src/app/services/category/category.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import { CategoryDetail } from '../../interfaces/categoryDetail';
import { CategoryShort } from '../../interfaces/categoryShort';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {

  URL: string = "http://localhost:8080/api/categories";

  constructor(private httpClient: HttpClient) {
  }

  public getAllCategories(): Observable<CategoryShort[]> {
    return this.httpClient.get<CategoryShort[]>(`${this.URL}`);
  }

  public getCategoryById(id: number): Observable<CategoryDetail> {
    return this.httpClient.get<CategoryDetail>((`${this.URL}/${id}`));
  }

  public addCategory(category: CategoryDetail): Observable<CategoryDetail> {
    return this.httpClient.post<CategoryDetail>(`${this.URL}`, category);
  }

  public updateCategory(category: CategoryDetail): Observable<CategoryDetail> {
    return this.httpClient.put<CategoryDetail>(`${this.URL}`, category);
  }

  public deleteCategory(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.URL}/${id}`);
  }

}

File: backend/src/main/java/org/yougrow/backend/services/CategoryService.java
package org.yougrow.backend.services;

import org.yougrow.backend.entities.Category;
import org.yougrow.backend.entities.projection.Category.CategoryDetail;
import org.yougrow.backend.entities.projection.Category.CategoryShort;

import java.util.List;
import java.util.Optional;

public interface CategoryService {

    List<CategoryShort> getAllCategoriesShort();

    Optional<CategoryDetail> getCategoryDetailById(Long id);

    Optional<Category> getCategoryById(Long id);

    Category addCategory(Category tag);

    Category updateCategory(Category tag);

    void deleteCategory(Long id);
}
